executor/src/witgen/jit/interpreter.rs

```diff
@@ -0,0 +1,546 @@
+use super::compiler::written_vars_in_effect;
+use super::effect::{Assertion, Effect};
+
+use super::symbolic_expression::{BinaryOperator, BitOperator, SymbolicExpression, UnaryOperator};
+use super::variable::{Cell, Variable};
+use crate::witgen::data_structures::finalizable_data::CompactDataRef;
+use crate::witgen::data_structures::mutable_state::MutableState;
+use crate::witgen::jit::effect::MachineCallArgument;
+use crate::witgen::machines::LookupCell;
+use crate::witgen::QueryCallback;
+use powdr_number::FieldElement;
+
+use std::collections::{BTreeSet, HashMap};
+
+// Witgen effects compiled into interpreter instructions.
+pub struct EffectsInterpreter<T: FieldElement> {
+    first_column_id: u64,
+    column_count: usize,
+    var_count: usize,
+    actions: Vec<InterpreterAction<T>>,
+}
+
+// Witgen effects compiled into "instructions".
+// Variables have been removed and replaced by their index in the variable list.
+enum InterpreterAction<T: FieldElement> {
+    ReadCell(usize, Cell),
+    ReadParam(usize, usize),
+    AssignExpression(usize, RPNExpression<T, usize>),
+    WriteCell(usize, Cell),
+    WriteParam(usize, usize),
+    WriteKnown(Cell),
+    MachineCall(u64, Vec<MachineCallArgumentIdx>),
+    Assertion(RPNExpression<T, usize>, RPNExpression<T, usize>, bool),
+}
+
+#[derive(Debug, Eq, PartialEq, Ord, PartialOrd)]
+pub enum MachineCallArgumentIdx {
+    // var index of the evaluated known argument expression
+    Known(usize),
+    // var index of the unknown
+    Unknown(usize),
+}
+
+impl<T: FieldElement> EffectsInterpreter<T> {
+    pub fn new(
+        first_column_id: u64,
+        column_count: usize,
+        known_inputs: &[Variable],
+        effects: &[Effect<T, Variable>],
+    ) -> Self {
+        let mut actions = vec![];
+        let mut var_mapper = VariableMapper::new();
+
+        Self::load_known_inputs(&mut var_mapper, &mut actions, known_inputs);
```

Comment:
Maybe these three should be put somewhere else, since they perform data type conversion instead of actual interpreting?

Reply:
I'd like to leave this as is for this PR, i have another branch where i did some optimizations and have a "Builder" struct for the interpreter